akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -59,38 +59,82 @@ private[akka] final case class TakeWhile[T](p: T ⇒ Boolean, decider: Supervisi
 /**
  * INTERNAL API
  */
-private[akka] final case class DropWhile[T](p: T ⇒ Boolean, decider: Supervision.Decider) extends PushStage[T, T] {
+private[stream] final case class DropWhile[T](p: T ⇒ Boolean) extends GraphStage[FlowShape[T, T]] {
+  val in = Inlet[T]("DropWhile.in")
+  val out = Outlet[T]("DropWhile.out")
+  override val shape = FlowShape(in, out)
+  override def initialAttributes: Attributes = DefaultAttributes.dropWhile
   var taking = false
 
-  override def onPush(elem: T, ctx: Context[T]): SyncDirective =
-    if (taking || !p(elem)) {
-      taking = true
-      ctx.push(elem)
-    } else {
-      ctx.pull()
+  def createLogic(inheritedAttributes: Attributes) = new SupervisionedGraphStageLogic(inheritedAttributes, shape) with InHandler with OutHandler {
+    override def onPush(): Unit = execute(() ⇒ {
+      val el = grab(in)
+      if (taking || !p(el)) {
+        taking = true
+        push(out, el)
+      } else
+        pull(in)
+    })
```

Comment:
Idea here is that `execute` mimic `try catch` so no code will be executed after exception anywhere inside parameter function.
Regarding `pulled twice` problem I'm not sure - perhaps it's inconsistent sate.  I would say that `pulled twice` is like run time exception and need to be fixed in opposite to regular exception that must be handled (according to Java vision at least). 
Let me know if there is some code snippet that shows different approach. Maybe I missed some important discussion on this topic.